### DISCUSSION TITLE
Does VPN leak other containers sharing network?

### DISCUSSION THREAD

**Discussion by duckduckbee** [haugene/docker-transmission-openvpn#2732] [Q&A]
I've been curious about how the killswitch works and I've been doing a bit of testing to see it in action. I've been running with a docker compose file similar to the example provided [here](https://haugene.github.io/docker-transmission-openvpn/tips-tricks/#example_jackett). I've been monitoring the containers from the host with wireshark to see what happens when the VPN dies. Perhaps the situation I've created is not realistic and isn't how the VPN is expected to fail but I've found when the openvpn process is killed within the container and the tunnel goes down (and container), there's a brief period of time in the jackett container where the network is still accessible but the tunnel is down and traffic is no longer going via VPN. The non tunnel routes still exist temporarily allowing traffic to go out. Is this behaviour expected given the way that it's failed, or is this potentially an issue? Or perhaps this behaviour is directly under the control of docker? I've ran some scenarios with iptables to stop outgoing traffic to the VPN to try and simulate a disconnect and that doesn't have the same issue, but I'm not sure if that's realistic either, if anyone has recommendations on creating a more realistic failure scenario like an actual disconnect I'd like to try it out. Thanks.

**Comment by KingKevin23** [comment]
I just ran into the same issue (at least if I understand you correctly). I was curious how the killswitch/binding thing works and tried to test it by killing the OpenVPN process and checking my connection afterwards:
```bash
docker exec transmission curl -sS "https://am.i.mullvad.net/connected"
sudo kill -9 <OpenVPN process id>
docker exec transmission curl -sS "https://am.i.mullvad.net/connected"
```
Rather surprising, my results showed an IP leak:
```
You are connected to Mullvad (server XXX). Your IP address is <VPN IP>
You are not connected to Mullvad. Your IP address is <REAL IP>
```
After some research I found #2626 and the `DROP_DEFAULT_ROUTE` option. Maybe I do not fully understand it yet, but this did not solve the problem either. So now I have basically the same questions: 

- Is this a realistic failure scenario? If no, how could I test this better?
- Is this expected behavior? If yes, can I change the setup in a way that stops this?

**Reply by duckduckbee** [reply]
Yes, you have understood me correctly, I could have been a bit clearer. That's essentially what I was doing, but I was doing it from within the jackett container. I didn't actually notice that it also happened with the transmission container too because the transmission container seems to die much quicker than the jackett container where there's a bigger window to see this issue. I've just reproduced what you've found too. I assume it's the same issue as what happens in the jackett container, the tunnel and its routes die leaving the non tunnel routes free to communicate without the VPN.

**Reply by duckduckbee** [reply]
I was unaware of the DROP_DEFAULT_ROUTE variable, so I've ran a couple of tests with that now. Like you, I can still reproduce the issue even without the default route, and have confirmed the default route is no longer there in the container.

It's a bit confusing because with just those routes its network access should be very limited, so not quite sure how it's happening now. I've tried manually deleting the tunnelled routes to try and get the same behaviour but get the expected Network is unreachable. I guess it's not as simple as what I originally suggested, something else must be happening. I'll see if I can work anything else out.

**Reply by duckduckbee** [reply]
OK, I caught it before the tunnel routes came back up again. In this window there is a brief period where the routes look like this:

```
default via 172.27.0.1 dev eth0
172.27.0.0/16 dev eth0 proto kernel scope link src 172.27.0.2
```

So it briefly regains a default route before it properly comes up again. Don't know enough at the moment to know why it does that, perhaps it's unique to us using `kill -9`, but would be good to know if this is a genuine issue or not. It doesn't seem beyond the realms of possibility that the openvpn process could die and cause this issue and leak traffic.

**Reply by KingKevin23** [reply]
Very interesting. The Transmission client should be safe as it should be bound directly to the VPN. For all other programs however, this could be an issue if this behavior can appear through other circumstances as well. As mentioned, a `kill -9` scenario is probably very unlikely

**Reply by duckduckbee** [reply]
Turns out we're restarting the container by killing the openvpn process. These two routes:
```
default via 172.27.0.1 dev eth0
172.27.0.0/16 dev eth0 proto kernel scope link src 172.27.0.2
```
Are just the initial state for the networking in the container, the same thing happens when bringing the container up in the first place. Maybe it's possible to configure the network to start from a more minimal network configuration? 

When the container restarts like this transmission won't be running in the first place, so that shouldn't matter. But I imagine that there is a problem for other containers using this network because they will just be exposed to that initial network state without warning. And it isn't just the `kill -9` that causes this it's anything that causes the container to restart.

**Reply by duckduckbee** [reply]
I didn't really get anywhere with trying to change network configuration get around this. 

I went back to the original example that I linked and noticed that it didn't have a restart policy for the service. I would assume the creators are aware that when the transmission container restarts that this small window exists, hence the lack of the restart policy. Requiring manual intervention for when the container goes down avoids this potential traffic leak assuming that you run `docker-compose down` to kill all other services and then `docker-compose up transmission` and then `docker-compose up` because the `depends_on` won't protect you by default from hitting the same small window with other services if you don't bring up transmission completely first. I doubt most users are aware of the specifics of this behaviour, I will try and raise an issue from this, if not to fix it, at least to increase awareness for others.

**Reply by edgd1er** [reply]
With kill -9 you are killing the kill switch wich is based on openvpn events.
On route down (connection lost), a script is executed. That script stops transmission.
```
--route-pre-down cmd
    Run command cmd before routes are removed upon disconnection.cmd consists of a path to script (or executable program), optionally followed by arguments. The path and arguments may be single- or double-quoted and/or escaped using a backslash, and should be separated by one or more spaces.
``` 
A second protection is dropping default route when vpn route is up. Whenever the vpn fails, traffic cannot be routed, should you have enabled that feature through env vars.

A simple kill (no option) <pid> should trigger the event, i did not test that though.
If you wish to test a disconnection , you may change the way openvpn is started to enable [management interface](https://openvpn.net/community-resources/management-interface/) , and then asks openvpn to disconnect. I expect the result to be the same as a kill cmd.

**Reply by duckduckbee** [reply]
Thanks for your response, there's a couple of things I had missed. That makes perfect sense that `kill -9` would skip the running of tunnelDown.sh and ends up not being a realistic scenario. I've tested `kill` as well now and can see the difference. 

It turns out that the majority of the problem described here isn't relevant to the killswitch, it's the default networking in the container prior to the configuring of the VPN and routes etc. Looking something like this:
```
default via 172.27.0.1 dev eth0
172.27.0.0/16 dev eth0 proto kernel scope link src 172.27.0.2
```
If the container is brought up with `docker-compose up` there's a chance that other services that `depends_on` it that share its network will have access to the default network provided by the container before the VPN has finished being set up. The same situation exists when the transmission-openvpn container restarts if a restart policy is present, services that share the network will be exposed to the initial docker container network configuration during its restart and can leak traffic. The killswitch is irrelevant here because it's on the start up of the container not downing of the container. Perhaps there's something else I'm missing, but is the intention not to have a restart policy for transmission-openvpn to prevent this from happening and require manual intervention to get around this?

**Reply by edgd1er** [reply]
You are right when another container is connecting to haugene's service, it can use the network whatever is the vpn status.
I suspect you may be able to lock the external routing while the vpn is not up, with iptables -P FORWARD DROP, and iptables -P FORWARD ACCEPT when up.
These command should be in start.sh to lock forward, in route-up to accept forward and route down script to drop forward.
I guess it would forbid external containers traffic to be routed throught this one.
It require to be tested though.

**Reply by duckduckbee** [reply]
That makes sense, although presumably there would still be a (smaller) window prior to running start.sh to add the iptables drop rule where the bridged default network would still be available to other containers?

I haven't looked into this in too much depth but could another potential option be reconfiguring the default bridge network using the `daemon.json` file like [this](https://docs.docker.com/network/drivers/bridge/#configure-the-default-bridge-network). Or perhaps using another network mode at first and then move to bridged after routes have been changed which could remove the window?

I have a feeling that generally users might assume that the other containers using transmission-openvpn will always be safe, similar to the behaviour with transmission. It might be worth making it clearer that this is not necessarily the case and there are other considerations that could be made to make it safer.
